File: Tracks/csharp/allergies/Allergies.cs
using System;
using System.Linq;

public enum Allergen
{
    Eggs = 1,
    Peanuts = 2,
    Shellfish = 4,
    Strawberries = 8,
    Tomatoes = 16,
    Chocolate = 32,
    Pollen = 64,
    Cats = 128
}

public class Allergies
{
    private readonly int mask;

    public Allergies(int mask) => this.mask = mask;

    public bool IsAllergicTo(Allergen allergen) => (mask & (int)allergen) > 0;

    public Allergen[] List() => (Enum.GetValues(typeof(Allergen)) as Allergen[]).Where(IsAllergicTo).ToArray();
}

File: Tracks/typescript/allergies/allergies.ts
const allergy = {
    eggs: 1,
    peanuts: 2,
    shellfish: 4,
    strawberries: 8,
    tomatoes: 16,
    chocolate: 32,
    pollen: 64,
    cats: 128
};

type AllergyOption = keyof typeof allergy;

export default class Allergies {
    private readonly allergyLevel: number;
    constructor(allergyLevel: number) {
        this.allergyLevel = allergyLevel;
    }

    allergicTo = (option: AllergyOption): boolean =>
        (this.allergyLevel & allergy[option]) === (allergy[option] | allergy[option]);

    list = (): string[] =>
        (Object.keys(allergy) as AllergyOption[]).reduce<string[]>(
            (list, allergyItem) => (this.allergicTo(allergyItem) ? [...list, allergyItem] : list),
            []
        );
}
